### ISSUE TITLE
escnn's conv, BN, relu is not equivariant?

### ISSUE THREAD

**Issue by Guptajakala** [QUVA-Lab/escnn#72]
I'm trying to train a image recognition network which should be invariant under SO(2). Here is a minimal example of my network that uses escnn's modules:
```
class EquivariantConvBN(enn.EquivariantModule):
  def __init__(self, in_type, out_type, kernel_size=3, stride=1, dilation=1, bias=True, use_norm=True):
    super().__init__()
    self.in_type = in_type
    self.out_type = out_type
    padding = (kernel_size - 1) // 2
    layers = []
    layers.append(enn.R2Conv(in_type, out_type, kernel_size=kernel_size, stride=stride, padding=padding, dilation=dilation, bias=bias, sigma=None))
    if use_norm:
      layers.append(enn.InnerBatchNorm(out_type))
    layers.append(enn.ReLU(out_type, inplace=True))

    self.layer = enn.SequentialModule(*layers)

  def forward(self, x):
    return self.layer(x)

  def evaluate_output_shape(self, input_shape):
    return self.layer.evaluate_output_shape(input_shape)

gspace = gspaces.rot2dOnR2(4)
n_feat = 16
in_type = enn.FieldType(gspace, [gspace.trivial_repr]*3)
out_type = enn.FieldType(gspace, [gspace.regular_repr]*n_feat)
conv = EquivariantConvBN(in_type, enn.FieldType(gspace, [gspace.regular_repr]*n_feat), kernel_size=7, stride=2).cuda().eval()
pool = nn.AdaptiveAvgPool2d(1).cuda()

rgb = np.random.uniform(0, 1, size=(480,480,3))   # Generate a random image
img = torch.tensor(rgb).float().cuda().permute(2,0,1)[None]
with torch.no_grad():
  out = pool(conv(enn.GeometricTensor(img, in_type)).tensor)
  for n_rot in range(0,3):
    rgb_tmp = cv2.rotate(rgb, n_rot)   # Rotate the image by 90/180/270 degs
    out_cur = pool(conv(enn.GeometricTensor(torch.tensor(rgb_tmp).float().cuda().permute(2,0,1)[None], in_type)).tensor)
    print(f'diff:{torch.abs(out-out_cur).max()}')
```
The output is 
```
diff:0.00496324896812439
diff:0.007862985134124756
diff:0.007146209478378296
```

The idea is to test under rotations of 90/180/270deg to the input image, whether the final output is the same. My understanding is that by using equivariant network, the output tensor should be the same under these rotations. However, I found the result is not. Is there anything I'm missing?

**Comment by Gabri95** [comment]
Hi @Guptajakala 

I suspect the problem is that you are testing the invariance of your model's output when the model's output is not invarint but equivariant.
Indeed, your output type contains n_feat copies of a regular representation, i.e. your 16*4 dimensional output splits in 16 blocks of size 4. The 4 channels within each block permute when the input rotates.
Your test is not accounting for this.

To properly check for equivariance, you should "rotated back" the output `out` by `n_rot`.
You could do that by wrapping `out` in a new GeometricTensor with type `out_type` and then use the `transform_fiber` method.
In other words, you should replace `out` with
```
GeometricTensor(out, out_type).transform_fibers(gspace.fibergroup.element(-n_rot)).tensor
```

This is a bit verbose since you unwrapped GeometricTensors and you used cv2 to rotate.
The code is a bit shorter if you use one of our pooling operators (which return GeometricTensors) and loop over `gspace.testing_elements()` (which returns already a list of GroupElements).

Hope this helps,
Gabriele

**Comment by Guptajakala** [comment]
@Gabri95 There is a pooling layer `AdaptiveAvgPool2d` at last. After the equivariance conv, suppose the feature shape is (B,D,H,W). After pooling, isn't it (B,D,1,1) and thus invariant? I guess even if I "rotate back", that single scalar in each HW dimension doesn't make any difference?

**Comment by Gabri95** [comment]
Hi @Guptajakala 

In this way, the output will be (approx) invariant to translations but not to rotations.
This is because the D channels in the output of shape (B, D,1,1) will rotate when the input rotates, since you chose `out_type = enn.FieldType(gspace, [gspace.regular_repr]*n_feat)`.
What you say would be correct if you used `out_type = enn.FieldType(gspace, [gspace.trivial_repr]*n_feat)`.

When using `out_type = enn.FieldType(gspace, [gspace.regular_repr]*n_feat)`, you can think of the channels dimension as being features over the rotation subgroup.
Check our [tutorial notebook](https://github.com/QUVA-Lab/escnn/blob/master/examples/introduction.ipynb) for a more intuitive description of the features of Steerable CNNs.

Hope this helps!
Gabriele
